### PR TITLE
1615 currency and subscription

### PR DESCRIPTION
**Pull request by xsalefter**
PR for #1615 for `currency` and `subscription` modules. Things that may need attention:

- [MultiValueMap](https://github.com/killbill/killbill/compare/work-for-release-0.23.x...xsalefter:1615-currency-and-subscription?expand=1#diff-98acdefe979d09a280923cfd61679af13f1593cec1f760c638055397a4145a94) and its [implementation](https://github.com/killbill/killbill/compare/work-for-release-0.23.x...xsalefter:1615-currency-and-subscription?expand=1#diff-6ad59a0d6aaf83d6e67fbec252f6a91b982e5d7cfbac23af147999cd81c4266d), especially the collection data type that will hold the values. Is `List` enough (at least for now)? Should we use `Set` instead of `List`? Do you think naming is Ok?

- [TenantCacheInvalidation](https://github.com/killbill/killbill/compare/work-for-release-0.23.x...xsalefter:1615-currency-and-subscription?expand=1#diff-50733d56a1fc0bcc660547aff8440dd1c1d969146a5ed7cf51e02924fc3f9a6c) use `MultiValueMap` instead of `Map<K, Collection<V>>`

- Some FIXME: [SubscriptionApiBase](https://github.com/killbill/killbill/compare/work-for-release-0.23.x...xsalefter:1615-currency-and-subscription?expand=1#diff-699f89a95013c79d65f3a2489c333b27c2dc0d176b9ef086afad26a71dbaf997R298), [DefaultSubscriptionInternalApi](https://github.com/killbill/killbill/compare/work-for-release-0.23.x...xsalefter:1615-currency-and-subscription?expand=1#diff-24d91c9d52326d020033a24837e591030b00dcd2dbf344e014445597a8e90b6aR186).

### CODE REVIEW

File: subscription/src/main/java/org/killbill/billing/subscription/api/SubscriptionApiBase.java [hunk 1]
```diff
@@ -289,18 +288,14 @@ protected SubscriptionBaseBundle createBundleForAccount(final UUID accountId,
...
-            }
-        }));
+        return internalSubscriptions.stream().map(this::createSubscriptionForApiUse).collect(Collectors.toUnmodifiableList());
     }
 
     protected DefaultSubscriptionBase createSubscriptionForApiUse(final SubscriptionBase internalSubscription) {
         return new DefaultSubscriptionBase((DefaultSubscriptionBase) internalSubscription, apiService, clock);
     }
 
+    // FIXME-1615 : Parameter InternalTenantContext context not used. Maybe worth to remove?
```
Comment (pierre) [inline subscription/src/main/java/org/killbill/billing/subscription/api/SubscriptionApiBase.java:298]: 👍

File: subscription/src/main/java/org/killbill/billing/subscription/api/svcs/DefaultSubscriptionInternalApi.java [hunk 2]
```diff
@@ -189,6 +185,7 @@ public DefaultSubscriptionBase apply(final SubscriptionBase subscriptionBase) {
     @Override
     public SubscriptionBaseBundle createBundleForAccount(final UUID accountId, final String bundleKey, final boolean renameCancelledBundleIfExist, final InternalCallContext context) throws SubscriptionBaseApiException {
         final DateTime now = context.getCreatedDate();
+        // FIXME-1615 : This is not used but intellij report there's possible side effect. But I don't think so. Thought?
```
Comment (pierre) [inline subscription/src/main/java/org/killbill/billing/subscription/api/svcs/DefaultSubscriptionInternalApi.java:188]: Yeah, I think it's removable.

**Review by pierre — APPROVED** [review]
Using `List` for `MultiValueMap` makes sense to me.

I've restarted CI, hopefully the previous failures were just flakiness.

File: subscription/src/main/java/org/killbill/billing/subscription/api/svcs/DefaultSubscriptionInternalApi.java (hunk 2 above)
Comment (xsalefter) [reply subscription/src/main/java/org/killbill/billing/subscription/api/svcs/DefaultSubscriptionInternalApi.java]: Done.

File: subscription/src/main/java/org/killbill/billing/subscription/api/SubscriptionApiBase.java [hunk 3]
```diff
@@ -78,7 +77,7 @@ public SubscriptionApiBase(final SubscriptionDao dao, final SubscriptionBaseApiS
     protected SubscriptionBaseBundle getActiveBundleForKey(final String bundleKey, final SubscriptionCatalog catalog, final InternalTenantContext context) throws CatalogApiException {
         final List<SubscriptionBaseBundle> existingBundles = dao.getSubscriptionBundlesForKey(bundleKey, context);
         for (final SubscriptionBaseBundle cur : existingBundles) {
-            final List<DefaultSubscriptionBase> subscriptions = dao.getSubscriptions(cur.getId(), ImmutableList.<SubscriptionBaseEvent>of(), catalog, context);
+            final List<DefaultSubscriptionBase> subscriptions = dao.getSubscriptions(cur.getId(), emptyList(), catalog, context);
```
Comment (sbrossie) [inline subscription/src/main/java/org/killbill/billing/subscription/api/SubscriptionApiBase.java:80]: I have seen other places where you don't add the static import and use `Collections.emptyList()` - we should try to be consistent, and I think not doing the static import is more readable

File: subscription/src/main/java/org/killbill/billing/subscription/api/SubscriptionApiBase.java [hunk 4]
```diff
@@ -116,7 +115,7 @@ protected SubscriptionBase getBaseSubscription(final UUID bundleId,
         if (result != null && !result.isEmpty()) {
             outputSubscriptions.addAll(result);
         }
-        Collections.sort(outputSubscriptions, DefaultSubscriptionInternalApi.SUBSCRIPTIONS_COMPARATOR);
+        sort(outputSubscriptions, DefaultSubscriptionInternalApi.SUBSCRIPTIONS_COMPARATOR);
```
Comment (sbrossie) [inline subscription/src/main/java/org/killbill/billing/subscription/api/SubscriptionApiBase.java:118]: Same remark on static import here.

File: util/src/test/java/org/killbill/billing/util/collect/TestMultiValueHashMap.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    public void putElement() {
+        final MultiValueMap<String, KeyValue> multiValueMap = new MultiValueHashMap<>();
+        multiValueMap.putElement("one", new KeyValue("a", "A"));
+        multiValueMap.putElement("two", new KeyValue("b", "B"), new KeyValue("c", "C"));
+
+        assertEquals(multiValueMap.size(), 2);
+        assertEquals(multiValueMap.get("one").size(), 1);
+        assertEquals(multiValueMap.get("two").size(), 2);
+
+        multiValueMap.putElement("two", new KeyValue("c", "C")); // same element, to prove that values size will keep increase.
```
Comment (sbrossie) [inline util/src/test/java/org/killbill/billing/util/collect/TestMultiValueHashMap.java:39]: So does it mean that the semantics is that we want to accept duplicates - i.e we end up with twice  the same `KeyValue("c", "C")` for the key `two` ?

Comment (xsalefter) [reply util/src/test/java/org/killbill/billing/util/collect/TestMultiValueHashMap.java:39]: @sbrossie Yes. What do you think? We can go like Guava did (like their `SetMultimap`, `ListMultiMap`, etc) but I don't think we don't need it now.

File: subscription/src/main/java/org/killbill/billing/subscription/api/SubscriptionApiBase.java (hunk 3 above)
Comment (xsalefter) [reply subscription/src/main/java/org/killbill/billing/subscription/api/SubscriptionApiBase.java]: Ok, lets use `Collections.emptyList()` and `Collections.sort()`

File: subscription/src/main/java/org/killbill/billing/subscription/api/SubscriptionApiBase.java (hunk 4 above)
Comment (xsalefter) [reply subscription/src/main/java/org/killbill/billing/subscription/api/SubscriptionApiBase.java]: fixed as well.

**Comment by xsalefter** [comment]
@pierre @sbrossie failed build is about 

`https://oss.sonatype.org/content/repositories/snapshots/org/kill-bill/billing/killbill-platform-osgi/0.41.0-f57a918-SNAPSHOT/killbill-platform-osgi-0.41.0-f57a918-20220325.085608-2.jar, status: 502 Bad Gateway`

Should I just commit something (empty space to file, etc) to re-run the build?

**Pull request merged**